nextcord/voicerec/core.py

```diff
@@ -0,0 +1,338 @@
+
+        # write in a loop as to avoid generating a huge memory buffer
+        while frames > 0:
+            if frames > 1000:
+                write(SILENCE_STRUCT_1000)
+                frames -= 1000
+            elif frames > 100:
+                write(SILENCE_STRUCT_100)
+                frames -= 100
+            elif frames > 10:
+                write(SILENCE_STRUCT_10)
+                frames -= 10
+            else:
+                write(SILENCE_STRUCT)
+                frames -= 1
+
+
+class TimeTracker:
+    def __init__(self) -> None:
+        self.starting_time: Optional[float] = clock_timestamp()  # Also used as filename
+        self.first_packet_time: Optional[float] = None
+        self.stopping_time: Optional[float] = None
+
+        self.users_times: dict[int, tuple[int, float]] = {}
+
+    def add_user(self, user, start_time) -> None:
+        if user not in self.users_times:
+            self.users_times[user] = start_time
+
+    def calculate_silence(self, user_id, timestamp, received_timestamp):
+        # process a packet for a registered user
+        if user_time := self.users_times.get(user_id):
+            delta_created_time = timestamp - user_time[0]
+            delta_received_time = (received_timestamp - user_time[1]) * AUDIO_HZ
+            difference = abs(100 - (delta_created_time * 100 / delta_received_time))
+
+            # calculate time since last audio packet
+            if difference > DIFFERENCE_THRESHOLD and delta_created_time != SILENCE_FRAME_SIZE:
+                silence = delta_received_time - SILENCE_FRAME_SIZE
+            else:
+                silence = delta_created_time - SILENCE_FRAME_SIZE
+
+        # first packet ever
+        elif not self.users_times:
+            # register first packet ever
+            self.first_packet_timestamp = received_timestamp
+            silence = 0
+
+        # first packet from user
+        else:
+            # calculate time since first packet from user
+            silence = (
+                (received_timestamp - self.first_packet_timestamp) * AUDIO_HZ
+            ) - SILENCE_FRAME_SIZE
+
+        # update receive times for next calculation
+        self.users_times[user_id] = (timestamp, received_timestamp)
+
+        return Silence(
+            frames=(max(0, int(silence / SILENCE_FRAME_SIZE)) * opus.DecoderThread.CHANNELS)
+        )
+
+
+class AudioWriter:
+    def __init__(self, temp_type: TempType, guild_id: int, user_id: int) -> None:
+        self.guild_id = guild_id
+        self.user_id = user_id
+
+        self.buffer: Union[BufferedWriter, BytesIO]
+        if temp_type == TempType.File:
+            self.buffer = open_tmp_file(guild_id, user_id, "wb+")
+        elif temp_type == TempType.Memory:
+            self.buffer = BytesIO()
+        else:
+            raise InvalidTempType(
+                f"Arg `temp_type` must be of type TempType not `{type(temp_type)}`"
+            )
+
+    def write(self, bytes) -> None:

```

Comment:
bytes should be `io.BufferedIOBase`

Reply:
(And I also shouldn't overwrite the type bytes, will resolve that next update

Reply:
closing, reversed change due to incorrect typing (missing methods used in io.BufferedIOBase